### PR TITLE
Fix Issue 17257 - Wrong recursive template destructor reflection

### PR DESCRIPTION
**Pull request by Kozzi11**

### CODE REVIEW

**Comment by japplegame** [comment]
Many thanks, Daniel.
But, unfortunately, it is need further development.
Slightly modify bugzilla sample:
```d
import std.string : format;

struct Bar(E) {
    void fun() {
        pragma(msg, format("fun:   Foo.__xdtor - %s", __traits(hasMember, E, "__xdtor")));
    }
    ~this() {
        pragma(msg, format("~this:   Foo.__xdtor - %s", __traits(hasMember, E, "__xdtor")));
    }
}

struct Foo {
    Bar!Foo foo;
    //~this() {}  <----- changed line (just remove destructor)
}
```
And we see wrong result again:
```
~this:   Foo.__xdtor - false
fun:   Foo.__xdtor - true
```

**Comment by Kozzi11** [comment]
@japplegame Ok I will look at it later today :)

**Comment by Kozzi11** [comment]
@japplegame OK, both cases should work now

File: src/ddmd/clone.d [hunk 1]
```diff
@@ -1008,6 +1010,13 @@ extern (C++) FuncDeclaration buildDtor(AggregateDeclaration ad, Scope* sc)
         auto sdv = (cast(TypeStruct)tv).sym;
         if (!sdv.dtor)
             continue;
+        // bug 17257
+        {
+            xdtor_fwd = sdv.dtor; // this dtor is temporaly it could be anything
```
Comment (andralex) [inline src/ddmd/clone.d]: s/temporaly /temporary, /

File: src/ddmd/clone.d [hunk 2]
```diff
@@ -1008,6 +1010,13 @@ extern (C++) FuncDeclaration buildDtor(AggregateDeclaration ad, Scope* sc)
...
+            auto _alias = new AliasDeclaration(Loc(), Id.__xdtor, xdtor_fwd);
+            _alias.addMember(sc, ad); // temporaly add to symbol table
```
Comment (andralex) [inline src/ddmd/clone.d]: s/temporaly/temporarily/

**Review by andralex — APPROVED** [review]
Good work, thx! cc @WalterBright

File: src/ddmd/clone.d [hunk 3]
```diff
@@ -1008,6 +1010,13 @@ extern (C++) FuncDeclaration buildDtor(AggregateDeclaration ad, Scope* sc)
         auto sdv = (cast(TypeStruct)tv).sym;
         if (!sdv.dtor)
             continue;
+        // bug 17257
```
Comment (WalterBright) [inline src/ddmd/clone.d]: Use links to bug reports, not just the bug number.

File: test/runnable/xdtor.d [hunk 4]
```diff
@@ -57,12 +57,33 @@ struct MixedDtor
...
+    ~this() @safe @nogc nothrow {}
+}
+
+struct Foo17257B
+{
+    Bar17257!Foo17257B foo;
+}
+
 static assert(!__traits(hasMember, SNoDtor, "__xdtor"));
 static assert(!__traits(hasMember, CNoDtor, "__xdtor"));
```
Comment (WalterBright) [inline test/runnable/xdtor.d]: Shouldn't the above two lines be above the added text?

File: test/runnable/xdtor.d [hunk 5]
```diff
@@ -57,12 +57,33 @@ struct MixedDtor
...
 
 void test2() @safe @nogc nothrow
 {
     FieldDtor a;
+    Foo17257A foo17257A;
+    Foo17257B foo17257B;
```
Comment (WalterBright) [inline test/runnable/xdtor.d]: This seems to be mixing together two test cases that have nothing to do with each other into one. Please make them separate test cases.

File: src/ddmd/clone.d [hunk 6]
```diff
@@ -1008,6 +1010,13 @@ extern (C++) FuncDeclaration buildDtor(AggregateDeclaration ad, Scope* sc)
...
+        {
+            xdtor_fwd = sdv.dtor; // this dtor is temporary it could be anything
+            auto _alias = new AliasDeclaration(Loc(), Id.__xdtor, xdtor_fwd);
```
Comment (WalterBright) [inline src/ddmd/clone.d]: Please don't use leading _ names.

File: src/ddmd/clone.d [hunk 7]
```diff
@@ -1008,6 +1010,13 @@ extern (C++) FuncDeclaration buildDtor(AggregateDeclaration ad, Scope* sc)
         auto sdv = (cast(TypeStruct)tv).sym;
         if (!sdv.dtor)
             continue;
+        // fix: https://issues.dlang.org/show_bug.cgi?id=17257
+        {
```
Comment (WalterBright) [inline src/ddmd/clone.d]: What are the braces here for? Some comment about what is going on would be nice.

File: test/runnable/xdtor.d [hunk 8]
```diff
@@ -79,8 +79,34 @@ void test2() @safe @nogc nothrow
...
+{
+    Bar17257!Foo17257B foo;
+}
+
+void test3() @safe @nogc nothrow
+{
+    Foo17257A foo17257A;
+    Foo17257B foo17257B;
+}
+
```
Comment (WalterBright) [inline test/runnable/xdtor.d]: As in the other test files with multiple independent tests, please delimit them with:
```
/**************************************/
... a test ...
/**************************************/
```

**Comment by japplegame** [comment]
This is really hard. Destructor detection works, but it doesn't fix https://forum.dlang.org/post/zpxzbctiijfhjujszvgl@forum.dlang.org
Also this code doesn't compile:
```d
import std.string : format;

struct Bar(E) {
	E* ptr;
    void fun() {
        pragma(msg, format("fun:   Foo.__xdtor - %s", __traits(hasMember, E, "__xdtor")));
    }
    void foo() {
        if(ptr !is null) ptr.__xdtor(); // works
    }
    ~this() {
        pragma(msg, format("~this:   Foo.__xdtor - %s", __traits(hasMember, E, "__xdtor")));
        if(ptr !is null) ptr.__xdtor(); // doesn't compile
    }
}

struct Foo {
    Bar!Foo foo;
}

void main() {}
```
With strange error:
```shell
test.d(13): Error: this for ~this needs to be type Bar not type Foo
```

**Comment by japplegame** [comment]
Awkward workaround is to save pointer to destruction function, then call it from real destructor.
```d
import std.stdio;
import std.experimental.allocator;
import std.experimental.allocator.mallocator : Mallocator;

struct Bar(E) {
    E* ptr = null;
    void function(E*) disposer;
    void allocate(int m) {
		ptr = Mallocator.instance.make!Foo(m);
		disposer = ptr => Mallocator.instance.dispose(ptr);
	}
    ~this() {
        if(ptr !is null) disposer(ptr);
    }
}

struct A {
    int m;
    ~this() { writefln("A.~this(%s)", m); }
}

struct Foo {
    A a;
    Bar!Foo bar;
    this(int m) { a = A(m); }
}

void main() {
    auto foo = Foo(1);
    foo.bar.allocate(10);
}
```

**Comment by Kozzi11** [comment]
@japplegame  'test.d(13): Error: this for ~this needs to be type Bar not type Foo' this is my fault, I will try to fix that

**Comment by benjones** [comment]
I just rebased this.  Can I push to this PR, or should I open a new one and reference this?

**Comment by thewilsonator** [comment]
If you have write permissions to the dmd repository, then yes (unless the author has explicitly disabled it for this PR, unlikely), otherwise no. If not ,please open a new PR to supersede this one.

**Pull request closed without merging**